src/Common/src/TypeSystem/Interop/InteropStateManager.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Internal.IL.Stubs;
+using Internal.TypeSystem.Interop;
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// This class manages and caches  interop state information
+    /// </summary>
+    public sealed class InteropStateManager
+    {
+        private readonly ModuleDesc _generatedAssembly;
+        private readonly NativeStructTypeHashtable _nativeStructHashtable;
+        private readonly StructMarshallingThunkHashTable _structMarshallingThunkHashtable;
+
+
+        public InteropStateManager(ModuleDesc generatedAssembly)
+        {
+            _generatedAssembly = generatedAssembly;
+            _structMarshallingThunkHashtable = new StructMarshallingThunkHashTable(this);
+            _nativeStructHashtable = new NativeStructTypeHashtable(this);
+        }
+
+        //
+        //  Struct Marshalling 
+        //  To support struct marshalling compiler needs to generate a native type which
+        //  imitates the original struct being passed to managed side with corresponding 
+        //  fields of marshalled types. Additionally it needs to generate three thunks
+        //      1. Managed to Native Thunk: For forward marshalling
+        //      2. Native to Managed Thunk: For reverse marshalling
+        //      3. Cleanup Thunk: for cleaning up any allocated resources
+        //
+        /// <summary>
+        /// Generates a Native struct type which imitates the managed struct
+        /// </summary>
+        internal NativeStructType GetStructMarshallingNativeType(TypeDesc managedType)
+        {
+            if (managedType is ByRefType)
+            {
+                managedType = ((ByRefType)managedType).GetParameterType();
+            }
+
+            Debug.Assert(managedType is MetadataType);
+
+            var typeKey = new NativeStructTypeHashtableKey((MetadataType)managedType, _generatedAssembly);
+            return _nativeStructHashtable.GetOrCreateValue(typeKey);
+        }
+
+        /// <summary>
+        ///  Generates a thunk to marshal the fields of the struct from managed to native
+        /// </summary>
+        internal MethodDesc GetStructMarshallingManagedToNativeThunk(TypeDesc managedType)
+        {
+            if (managedType is ByRefType)
+            {
+                managedType = ((ByRefType)managedType).GetParameterType();
```

Comment:
You can now call GetParameterType on TypeDesc without casting to ByRef type. It does the same thing but simplifies the code a bit.